### DISCUSSION TITLE
Help compiling Win10 & Generating VS files.

### DISCUSSION THREAD

**Discussion by danielmilvus** [FreeRDP/FreeRDP#7002] [Q&A]
Hello, 
I´m trying to compile FreeRDP on Windows 10 to integrate a project and I´m getting errors I couldn´t solve by myself.

I could compile all the dependencies following the instructions but I get error trying to compile FreeRDP itself.

Attached is the output from this command: (Build Output.txt)
cmake -GNinja -BC:\Users\Milvus\source\repos\freerdp\bin -HC:\Users\Milvus\source\repos\freerdp -DCMAKE_INSTALL_PREFIX=C:\Users\Milvus\source\repos\freerdp\bin\release -DCMAKE_BUILD_TYPE=Release
[CMakeError.log](https://github.com/FreeRDP/FreeRDP/files/6452343/CMakeError.log)
[CMakeOutput.log](https://github.com/FreeRDP/FreeRDP/files/6452344/CMakeOutput.log)
[Build Output.txt](https://github.com/FreeRDP/FreeRDP/files/6452346/Build.Output.txt)

I´ve also tried to generate the project to buil in VS2019 but also got error, output files are attached too. (CMakeError & CMakeOutput)

Could you help me ?

My best 
Daniel.

**Comment by hardening** [comment]
Try to disable ` libusb`  support.

**Reply by danielmilvus** [reply]
Hello @hardening thanks for helping...
How can I do it ?

**Reply by hardening** [reply]
run ` cmake-gui`  in the build directory and adjust build options

**Reply by danielmilvus** [reply]
@hardening After doing it, I get the same output as when I try to generate Visual Studio project following these instructions:
https://github.com/FreeRDP/FreeRDP/wiki/Build-on-Windows-with-Visual-C---2010

Output attached.

Thanks again for helping, appreciate !
[CMakeError.log](https://github.com/FreeRDP/FreeRDP/files/6453263/CMakeError.log)
[CMakeOutput.log](https://github.com/FreeRDP/FreeRDP/files/6453264/CMakeOutput.log)

**Reply by hardening** [reply]
And what happens when you ninja after that ?

**Reply by danielmilvus** [reply]
Using the command: 'cmake -GNinja -BC:\Users\Milvus\source\repos\freerdp -HC:\Users\Milvus\source\repos\freerdp -DCMAKE_INSTALL_PREFIX=C:\Users\Milvus\source\repos\freerdp -DCMAKE_BUILD_TYPE=Release'

I have the same output error I´ve mentioned before!
I could check I have the files the compiler could not find under my 'C:\Strawberry\c\i686-w64-mingw32\include'
Output files attached !

Please, if you could help somehow, could be a remote session, I´ve lost two entire days in the same issue, I tried to start all over again twice, got to the same point.
I can offer you a paid hour to fix my environment if needed.

[CMakeError.log](https://github.com/FreeRDP/FreeRDP/files/6454967/CMakeError.log)
[CMakeOutput.log](https://github.com/FreeRDP/FreeRDP/files/6454968/CMakeOutput.log)

**Reply by danielmilvus** [reply]
Here goes the 'CMake .' output :

C:\Users\Milvus\source\repos\freerdp>cmake .
-- Building for: Visual Studio 16 2019
CMake Deprecation Warning at CMakeLists.txt:21 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19042.
-- The C compiler identification is MSVC 19.28.29914.0
-- The CXX compiler identification is MSVC 19.28.29914.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: C:/Strawberry/perl/bin/pkg-config.bat (found version "0.26")
FREERDP_VERSION=2.3.2
-- Git Revision 2.3.2
-- MSVC: using dynamically-linked runtime (/MD and /MDd).
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
-- Looking for include file fcntl.h
-- Looking for include file fcntl.h - found
-- Looking for include file unistd.h
-- Looking for include file unistd.h - not found
-- Looking for include file execinfo.h
-- Looking for include file execinfo.h - not found
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Looking for include file sys/modem.h
-- Looking for include file sys/modem.h - not found
-- Looking for include file sys/filio.h
-- Looking for include file sys/filio.h - not found
-- Looking for include file sys/sockio.h
-- Looking for include file sys/sockio.h - not found
-- Looking for include file sys/strtio.h
-- Looking for include file sys/strtio.h - not found
-- Looking for include file sys/select.h
-- Looking for include file sys/select.h - not found
-- Looking for include file syslog.h
-- Looking for include file syslog.h - not found
-- Performing Test HAVE_TM_GMTOFF
-- Performing Test HAVE_TM_GMTOFF - Failed
-- Skipping disabled feature X11 for X11 (X11 client and server)
-- Skipping disabled feature Wayland for Wayland (Wayland client)
-- Skipping disabled feature ZLIB for compression (data compression)
-- Finding required feature OpenSSL for cryptography (encryption, certificate validation, hashing functions)
CMake Error at cmake/FindOpenSSL.cmake:316 (list):
  list GET given empty list
Call Stack (most recent call first):
  cmake/FindFeature.cmake:16 (find_package)
  CMakeLists.txt:820 (find_feature)


CMake Error at cmake/FindOpenSSL.cmake:317 (list):
  list GET given empty list
Call Stack (most recent call first):
  cmake/FindFeature.cmake:16 (find_package)
  CMakeLists.txt:820 (find_feature)


CMake Error at cmake/FindOpenSSL.cmake:319 (list):
  list GET given empty list
Call Stack (most recent call first):
  cmake/FindFeature.cmake:16 (find_package)
  CMakeLists.txt:820 (find_feature)


CMake Error at cmake/FindOpenSSL.cmake:321 (list):
  list GET given empty list
Call Stack (most recent call first):
  cmake/FindFeature.cmake:16 (find_package)
  CMakeLists.txt:820 (find_feature)


-- Found OpenSSL: C:/Users/Milvus/source/repos/openssl/lib/libssl.lib;C:/Users/Milvus/source/repos/openssl/lib/libcrypto.lib (found version ".0.0`")
-- Skipping optional feature MbedTLS for cryptography (encryption, certificate validation, hashing functions)
--     Enable feature MbedTLS using "-DWITH_MBEDTLS=ON"
-- Skipping disabled feature OpenSLES for multimedia (OpenSLES audio / video)
-- Skipping disabled feature OSS for sound (audio input, audio output and multimedia redirection)
-- Skipping disabled feature ALSA for sound (audio input, audio output and multimedia redirection)
-- Skipping disabled feature Pulse for sound (audio input, audio output and multimedia redirection)
-- Skipping disabled feature Cups for printing (printer device redirection)
-- Skipping disabled feature PCSC for smart card (smart card device redirection)
-- Skipping disabled feature FFmpeg for multimedia (multimedia redirection, audio and video playback)
-- Skipping optional feature JPEG for codec (use JPEG library)
--     Enable feature JPEG using "-DWITH_JPEG=ON"
-- Skipping optional feature x264 for codec (use x264 library)
--     Enable feature x264 using "-DWITH_X264=ON"
-- Skipping optional feature OpenH264 for codec (use OpenH264 library)
--     Enable feature OpenH264 using "-DWITH_OPENH264=ON"
-- Skipping optional feature OpenCL for codec ([experimental] use OpenCL library)
--     Enable feature OpenCL using "-DWITH_OPENCL=ON"
-- Skipping optional feature GSM for codec (GSM audio codec library)
--     Enable feature GSM using "-DWITH_GSM=ON"
-- Skipping optional feature LAME for codec (lame MP3 audio codec library)
--     Enable feature LAME using "-DWITH_LAME=ON"
-- Skipping optional feature FAAD2 for codec (FAAD2 AAC audio codec library)
--     Enable feature FAAD2 using "-DWITH_FAAD2=ON"
-- Skipping optional feature FAAC for codec ([experimental] FAAC AAC audio codec library)
--     Enable feature FAAC using "-DWITH_FAAC=ON"
-- Skipping optional feature soxr for codec (SOX audio resample library)
--     Enable feature soxr using "-DWITH_SOXR=ON"
-- Skipping optional feature GSSAPI for auth ([experimental] add kerberos support)
--     Enable feature GSSAPI using "-DWITH_GSSAPI=ON"
-- Skipping optional feature IPP for performance (Intel Integrated Performance Primitives library)
--     Enable feature IPP using "-DWITH_IPP=ON"
-- Using OpenSSL Version: .0.0`
using default plugins location
CMake Deprecation Warning at winpr/CMakeLists.txt:18 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Looking for include file stdbool.h
-- Looking for include file stdbool.h - found
-- Looking for include file stdint.h
-- Looking for include file stdint.h - found
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Looking for timer_create
-- Looking for timer_create - not found
-- Looking for timer_delete
-- Looking for timer_delete - not found
-- Looking for timer_settime
-- Looking for timer_settime - not found
-- Looking for timer_gettime
-- Looking for timer_gettime - not found
-- Checking for module 'cairo'
--   Can't find cairo.pc in any of C:/Users/Milvus/source/repos/gtk+-dev-2.8.9/lib/pkgconfig C:/strawberry/c/lib/pkgconfig C:/Strawberry/c/lib/pkgconfig
use the PKG_CONFIG_PATH environment variable, or
specify extra search paths via 'search_paths'
-- Could NOT find Cairo (missing: CAIRO_LIBRARIES CAIRO_INCLUDE_DIR)
CMake Warning at libfreerdp/CMakeLists.txt:102 (message):
  -DWITH_SWSCALE=OFF and -DWITH_CAIRO=OFF, compiling without image scaling
  support!


CMake Deprecation Warning at rdtk/CMakeLists.txt:18 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Adding STATIC channel client server "drdynvc": Dynamic Virtual Channel Extension
-- Adding DYNAMIC channel client "video": Video optimized remoting Virtual Channel Extension
-- Adding DYNAMIC channel client "urbdrc": USB Devices Virtual Channel Extension
-- Adding DEVICE channel client "smartcard": Smart Card Virtual Channel Extension
-- Adding STATIC channel client server "remdesk": Remote Assistance Virtual Channel Extension
-- Adding STATIC channel client server "rdpsnd": Audio Output Virtual Channel Extension
-- Adding DYNAMIC channel client "rdpgfx": Graphics Pipeline Extension
-- Adding DYNAMIC channel client "rdpei": Input Virtual Channel Extension
-- Adding STATIC channel client server "rdpdr": Device Redirection Virtual Channel Extension
-- Adding STATIC channel client "rdp2tcp": Tunneling TCP over RDP
-- Adding STATIC channel client "rail": Remote Programs Virtual Channel Extension
-- Adding DEVICE channel client "printer": Print Virtual Channel Extension
-- Adding DYNAMIC channel client "geometry": Geometry tracking Virtual Channel Extension
-- Adding STATIC channel client server "encomsp": Multiparty Virtual Channel Extension
-- Adding DYNAMIC channel client server "echo": Echo Virtual Channel Extension
-- Adding DEVICE channel client "drive": Drive Redirection Virtual Channel Extension
-- Adding DYNAMIC channel client "disp": Display Update Virtual Channel Extension
-- Adding STATIC channel client server "cliprdr": Clipboard Virtual Channel Extension
-- Adding DYNAMIC channel client server "audin": Audio Input Redirection Virtual Channel Extension
CMake Warning (dev) at C:/Program Files/CMake/share/cmake-3.20/Modules/CPack.cmake:690 (option):
  Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
  --help-policy CMP0077" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  For compatibility with older versions of CMake, option is clearing the
  normal variable 'CPACK_BINARY_ZIP'.
Call Stack (most recent call first):
  CMakeCPack.cmake:102 (include)
  CMakeLists.txt:1096 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring incomplete, errors occurred!
See also "C:/Users/Milvus/source/repos/freerdp/CMakeFiles/CMakeOutput.log".
See also "C:/Users/Milvus/source/repos/freerdp/CMakeFiles/CMakeError.log".

**Reply by akallabeth** [reply]
I´ve updated the wiki at https://github.com/FreeRDP/FreeRDP/wiki/Build-on-Windows-with-Visual-C to reflect some more recent changes.
As for your build problem, the OpenSSL detected is broken.

**Reply by danielmilvus** [reply]
Hello @akallabeth nice to meet you!

That´s weird, I´ve downloaded and builded OpenSSL brom GIT!
After changing to the installed version I could generate the solution and compile it ! 
(Would be nice to know how did you figured it out)

Is there any documentation on testing ? ( I´m trying to run from VS but I get "Unable to start, system cannot find ALL_BUILD file specified."
I see you have build options for 'make test' but I could not accomplish to test.
The 'wfreerdp.exe' generated by compiling does not open, nor give any error.

My goal is to embbed  FreeRDP into my SW, if you can help with any tips or best ways to go, I can produce documentation over it and make it available for your wiki !
If FreeRDP Co might be interested in collaborate to this project, we can talk further !

 Thank you very much for helping, I really appreciate a lot !
My best regards.
Dan

**Reply by danielmilvus** [reply]
Hello Mrs @akallabeth & @hardening ,
I can´t thank you enough for your patience and help !!!
Sometimes we are so focused on make it work, we can´t avoid to loose something along the way!
I could compile and test Win-Win, now I´ll start to work on it !!

Though I could not compile with X11.
Could you kindly help me on this last step ?
Attached you will find my build output and cmake outputs.

I could check it complains about rt.lib: LINK : fatal error LNK1104: cannot open file 'rt.lib'
Also it could not find XShm library.

I could try to find the lib, install and so on, but I know this probably isn´t the right way to go. (By experience from the last issue I had).

Appreciate a lot your suport !!!
Regards !
Daniel.
[CMakeError.log](https://github.com/FreeRDP/FreeRDP/files/6462344/CMakeError.log)
[CMakeOutput.log](https://github.com/FreeRDP/FreeRDP/files/6462345/CMakeOutput.log)
[Build_Output.txt](https://github.com/FreeRDP/FreeRDP/files/6462346/Build_Output.txt)

**Comment by akallabeth** [comment]
@danielmilvus what are you doing? `X11` is a linux backend, you´re trying to build on windows, that will never work...
just stick with the default options for now and get something that builds before you randomly force activate stuff you don´t quite understand yet ;)

**Reply by danielmilvus** [reply]
Good morning @akallabeth !
I´m doing great, how about you ?

I just tought I need to compile with X11 somehow (Some win version of it) in order to RDP Win-Linux !
Thanks anyway, have a wonderfull day !

**Reply by akallabeth** [reply]
@danielmilvus oh, good morning :)
no, there are platform specific clients.
